File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
class Solution {
    int[] dx = new int[]{0,1,0,-1};
    int[] dy = new int[]{1,0,-1,0};
    int N = 0;
    int M = 0;
    public int solution(int[][] land) {
        N = land.length;
        M = land[0].length;
        int[] answers = new int[M];
        
        for(int i = 0; i < N ;i++)
            for(int j = 0; j< M; j++)
                if(land[i][j] == 1)
                    bfs(land, answers, j, i);
        int answer = 0;
        for(int itm:answers)
            answer = Math.max(itm, answer);
        
        return answer;
    }
    private void bfs(int[][] land, int[] answers, int sx, int sy){
        Deque<int[]> dq = new ArrayDeque<>();
        land[sy][sx] = 0;
        dq.addFirst(new int[]{sx, sy});
        int mn = sx;
        int mx = sx;
        int cnt = 1;
        while(!dq.isEmpty()){
            int[] xy = dq.pollFirst();
            for(int i = 0; i < 4; i++){
                int nx = xy[0]+dx[i];
                int ny = xy[1]+dy[i];
                
                if(nx<0 || ny <0 || M<=nx || N<=ny)
                    continue;
                if(land[ny][nx] == 1){
                    cnt++;
                    land[ny][nx] = 0;
                    mx = Math.max(nx,mx);
                    mn = Math.min(nx,mn);
                    dq.addLast(new int[]{nx,ny});
                }
            }
        }
        
        for(int i = mn; i<=mx;i++)
            answers[i] += cnt;
    }
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import defaultdict
import sys
sys.setrecursionlimit(10**6)

def solution(land):
    N = len(land)
    M = len(land[0])
    num = 2
    dd = defaultdict(int)
    
    dx = [0,1,0,-1]
    dy = [1,0,-1,0]
    
    def dfs(n, x, y):
        dd[n] += 1
        land[y][x] = n
        for i in range(4):
            nx, ny = x+dx[i],y+dy[i]
            if 0 <= nx < M and 0 <= ny < N and land[ny][nx] == 1:
                dfs(n, nx, ny)
    
    for i in range(N):
        for j in range(M):
            if land[i][j] == 1:
                cnt = 0
                dfs(num,j,i)
                num += 1
    ans = 0
    for x in range(M):
        sett = {land[y][x] for y in range(N)}
        tmp = [dd[itm] for itm in sett]
        ans = max(sum(tmp), ans)
        
    return ans